evac/worker.py

```diff
@@ -723,21 +724,22 @@ def do_simulation(self):
         self.wlogger.debug('Final results gathered: {}'.format(self.cross_building_results))
 
     def change_pynavmesh_due_to_smoke(self):
-        critical_conditions_rooms = {}
-        for f in self.floors:
-            critical_conditions_rooms[f.floor] = []
-        for floor, rooms_smoke in self.smoke_opacity[-1].items():
-            for room_id, room_od in rooms_smoke.items():
-                    if room_od > 0.666:
-                        critical_conditions_rooms[self.rooms[room_id]['floor']].append(room_id)
+        # use floor parameter instead of reading opacity from smoke_opacity
+        # avoid recreating pynavmesh if not necessary
         for floor in self.floors:
-            if len(critical_conditions_rooms[floor.floor]) > 0:
-                self.generate_new_pynavmesh(floor, critical_conditions_rooms[floor.floor])
+            try:
+                self.previous_critical_rooms[floor.floor]
+            except KeyError:
+                self.previous_critical_rooms[floor.floor] = []
+            if floor.unavailable_rooms != self.previous_critical_rooms[floor.floor]:
+                if len(floor.unavailable_rooms) > 0:
```

Comment:
Probably those conditions should be nested the other way around.
```suggestion

            if len(floor.unavailable_rooms) > 0:
                if floor.unavailable_rooms != self.previous_critical_rooms[floor.floor]:
```